### PR TITLE
[compiler] Add `If` and `MakeStruct` Simplifications

### PR DESCRIPTION
**Pull request by ehigham**
Add the following transformations:
```scala
MakeStruct("a" -> GetField(o, "x"), ...) -> CastRename(SelectFields(o, ["x"..]), newtype)
If(IsNA(x), NA(x.typ), x) -> x
```

The changes to `SStructView` (nee `SSubsetStruct`) were as a result of a bud that prevented subsetting and then renaming to an excluded field, ie `{x, y, z} subset {z} rename {x}`. Now `SStructView` leaves its parent `SType` unmodified and casts loads through the parent to the appropriate `SValue`.

### CODE REVIEW

**Review by patrick-schultz — DISMISSED** [review]
nice!

**Review by patrick-schultz — APPROVED** [review]
looks good

**Comment by ehigham** [comment]
Merging in https://github.com/hail-is/hail/pull/14233 causes the failure in `test_union_rows1`. Some strangeness with these new dependencies - running without this commit and everything works fine.

File: hail/src/main/scala/is/hail/types/physical/stypes/concrete/SSubsetStruct.scala [hunk 1]
```diff
@@ -6,47 +6,59 @@ import is.hail.expr.ir.{EmitCodeBuilder, IEmitCode}
...
+    // note we may have subsetted a parent struct{x,y,z} to struct{z} then renamed to struct{x}
+    // must only tell parent to castRename what it knows as `z`, leaving others intact
+    val newParent = parent.castRename(
+      TStruct(parent.virtualType.fields.zipWithIndex.map { case (f, i) =>
+        oldToNewFieldMapping.get(i) match {
+          case Some(idx) => f.copy(typ = newVirtualType.types(idx))
+          case None => f
+        }
+      })
+    )
```
Comment (patrick-schultz) [inline hail/src/main/scala/is/hail/types/physical/stypes/concrete/SSubsetStruct.scala:54]: Do we need this? I thought the point of defining equality to ignore field names in the parent was to not have to rename the parent at all.

File: hail/src/main/scala/is/hail/types/physical/stypes/concrete/SSubsetStruct.scala [hunk 2]
```diff
@@ -6,47 +6,59 @@ import is.hail.expr.ir.{EmitCodeBuilder, IEmitCode}
...
+
+    new SSubsetStruct(newParent.asInstanceOf[SBaseStruct], newVirtualType.fieldNames) {
+      override lazy val newToOldFieldMapping: Map[Int, Int] =
+        SSubsetStruct.this.newToOldFieldMapping
+      override lazy val virtualType: TStruct =
+        newVirtualType
+    }
```
Comment (patrick-schultz) [inline hail/src/main/scala/is/hail/types/physical/stypes/concrete/SSubsetStruct.scala:61]: I don't like the anonymous subclassing here. It makes it difficult to say what an `SSubsetStruct` really is. I think the core issue forcing the need subclass is that `newToOldFieldMapping` should really be a constructor argument (as also evidenced by its appearence in the definition of equality). Then there should be no need to subclass.

Concretely, I suggest making the default constructor
```
class SSubsetStruct(
  private val parent: SBaseStruct,
  private val fieldNames: IndexedSeq[String],
  private val oldToNewFieldMapping
) extends SBaseStruct { ...
```
and adding a convenience constructor taking only the first two, and computing a default `oldToNewFieldMapping` as we do now.

**Review by patrick-schultz — CHANGES_REQUESTED** [review]

File: hail/src/main/scala/is/hail/types/physical/stypes/concrete/SSubsetStruct.scala (hunk 2 above)
Comment (ehigham) [reply hail/src/main/scala/is/hail/types/physical/stypes/concrete/SSubsetStruct.scala:61]: Sure - that sounds like a better change.

File: hail/src/main/scala/is/hail/types/physical/stypes/concrete/SSubsetStruct.scala (hunk 1 above)
Comment (ehigham) [reply hail/src/main/scala/is/hail/types/physical/stypes/concrete/SSubsetStruct.scala]: No longer needed

**Comment by ehigham** [comment]
> Merging in #14233 causes the failure in `test_union_rows1`. Some strangeness with these new dependencies - running without this commit and everything works fine.

Doesn't seem to be an issue anymore...

File: hail/src/main/scala/is/hail/types/physical/stypes/concrete/SStructView.scala [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+  lazy val newToOldFieldMapping: Map[Int, Int] =
+    restrict.fields.view.map(f => f.index -> parent.fieldIdx(f.name)).toMap
+
+  override lazy val fieldTypes: IndexedSeq[SType] =
+    Array.tabulate(size) { i =>
+      parent
+        .fieldTypes(newToOldFieldMapping(i))
+        .castRename(rename.fields(i).typ)
+    }
```
[the remark this answers is not in the document]
Comment (ehigham) [reply hail/src/main/scala/is/hail/types/physical/stypes/concrete/SStructView.scala:50]: OK

**Review by patrick-schultz — APPROVED** [review]

**Pull request merged**